File: crates/pylon-server/src/workers.rs
//
// This source file is part of the Pylon open source project.
//
// Copyright (c) 2026 Jaldis B.V.
//
// Licensed under the MIT OR Apache-2.0 license (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://opensource.org/licenses/MIT
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

//! Background index/cache workers — spawns `pylon_workers`' native
//! LISTEN/NOTIFY loops as Tokio tasks on server startup, mirroring
//! `pylon/cli/commands/worker.py::build_worker_tasks(schema, config,
//! shared_cache=True)`'s own worker set (`asyncio.ensure_future`'d there,
//! `tokio::spawn`'d here — cancelled on shutdown either way).
//!
//! The signals dispatcher is deliberately excluded here (unlike every
//! other worker `build_worker_tasks` can return) — it needs a live
//! reference to each `@pylon.signal`-decorated Python callable, which only
//! exists in a Python process. A user relying on signals runs `pylon
//! worker start` in its own process alongside `pylon-server` — there's no
//! way for this binary to launch it in-process the way the old Python ASGI
//! server once did.

use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use pylon_core::schema::{SchemaDescriptor, SearchBackend};
use pylon_workers::{
    CacheInvalidationWorker, MeilisearchClient, OpenSearchClient, ProviderConfig, SearchIndexWorker, VectorIndexWorker,
};

use crate::config::{ApiStyle, Config};

/// Matches `build_worker_tasks`'s own defaults (`pylon/cli/commands/
/// worker.py`) — `pylon serve` never overrides these, only the `worker
/// start` CLI's `--batch-size`/`--poll-interval` flags do.
const BATCH_SIZE: i64 = 50;
const POLL_INTERVAL: Duration = Duration::from_secs(30);
/// Matches `run_meilisearch_worker`/`run_opensearch_worker`'s own default
/// `timeout_secs` (`pylon-py/src/workers.rs`).
const SEARCH_TIMEOUT: Duration = Duration::from_secs(10);

/// Per-worker override, set via `pylon-server`'s `--<worker>-worker` /
/// `--disable-<worker>-worker` flag pairs — `None` (the default) defers to
/// whatever `schema`/`config` would otherwise imply; `Some(false)` always
/// wins and skips the worker entirely, for deployments that run it in its
/// own process (`pylon worker start`) instead and don't want `pylon-server`
/// double-spawning it in-process. `Some(true)` explicitly opts back in —
/// for the vector/search workers this only matters if they'd otherwise be
/// skipped for lack of a matching schema declaration, since there's
/// nothing else gating them; it can't invent vector/search indexes the
/// schema doesn't have. For the cache worker, `Some(true)` still can't
/// start it if `[cache].enabled = false`, since no LMDB handle was ever
/// opened for it to attach to (that handle is shared with every `Client`'s
/// own read-through query cache in `AppState`, so forcing it open here
/// would silently turn query caching on for every request too — a bigger
/// change than "just this worker" — see `main.rs`'s CLI help text).
#[derive(Default, Clone, Copy)]
pub struct WorkerToggles {
    pub vector: Option<bool>,
    pub search: Option<bool>,
    pub cache: Option<bool>,
}

/// Spawns whatever background workers `schema`/`config` imply as detached
/// Tokio tasks, returning their `JoinHandle`s so the caller can abort them
/// on shutdown. A worker that fails to start (missing config, bad HTTP
/// client construction) logs and is simply skipped — mirrors
/// `build_worker_tasks`'s own `log.warning(...); continue` pattern, not a
/// startup-fatal condition (unlike the base DB connection itself).
pub fn spawn(
    schema: &SchemaDescriptor,
    config: &Config,
    dsn: &str,
    cache: Option<Arc<pylon_cache::Cache>>,
    toggles: WorkerToggles,
) -> Vec<tokio::task::JoinHandle<()>> {
    let mut handles = Vec::new();

    if toggles.vector == Some(false) {
        eprintln!("pylon-server: VectorIndexWorker disabled (--disable-vector-worker)");
    } else {
        let providers = build_providers(schema, config);
        if !providers.is_empty() {
            match VectorIndexWorker::new(schema.clone(), providers) {
                Ok(worker) => {
                    eprintln!("pylon-server: VectorIndexWorker started");
                    handles.push(spawn_index_worker(dsn.to_string(), worker, "VectorIndexWorker"));
                }
                Err(e) => eprintln!("pylon-server: failed to start VectorIndexWorker: {e}"),
            }
        }
    }

    if toggles.search == Some(false) {
        eprintln!("pylon-server: search index workers disabled (--disable-search-worker)");
    } else {
        let want_opensearch = schema.types.iter().any(|t| {
            t.search_indexes
                .iter()
                .any(|si| si.backend == SearchBackend::OpenSearch)
        });
        let want_meilisearch = schema.types.iter().any(|t| {
            t.search_indexes
                .iter()
                .any(|si| si.backend == SearchBackend::Meilisearch)
        });

        if want_opensearch {
            match config.search.get("default") {
                None => eprintln!(
                    "pylon-server: SearchIndex(backend=OpenSearch) declared but no [search] config found; skipping"
                ),
                Some(search_cfg) => {
                    let base_url = format!("http://{}:{}", search_cfg.host, search_cfg.port);
                    let auth = search_cfg.user.as_deref().zip(search_cfg.password.as_deref());
                    match OpenSearchClient::new(&base_url, auth, SEARCH_TIMEOUT) {
                        Ok(client) => {
                            eprintln!("pylon-server: OpenSearchWorker started  base_url={base_url}");
                            let worker = SearchIndexWorker::new(schema.clone(), client, "OpenSearch");
                            handles.push(spawn_index_worker(dsn.to_string(), worker, "OpenSearchWorker"));
                        }
                        Err(e) => eprintln!("pylon-server: failed to start OpenSearchWorker: {e}"),
                    }
                }
            }
        }

        if want_meilisearch {
            match config.search.get("default") {
                None => eprintln!(
                    "pylon-server: SearchIndex(backend=Meilisearch) declared but no [search] config found; skipping"
                ),
                Some(search_cfg) => {
                    let base_url = format!("http://{}:{}", search_cfg.host, search_cfg.port);
                    match MeilisearchClient::new(&base_url, search_cfg.api_key.as_deref(), SEARCH_TIMEOUT) {
                        Ok(client) => {
                            eprintln!("pylon-server: MeilisearchWorker started  base_url={base_url}");
                            let worker = SearchIndexWorker::new(schema.clone(), client, "Meilisearch");
                            handles.push(spawn_index_worker(dsn.to_string(), worker, "MeilisearchWorker"));
                        }
                        Err(e) => eprintln!("pylon-server: failed to start MeilisearchWorker: {e}"),
                    }
                }
            }
        }
    }

    if toggles.cache == Some(false) {
        if cache.is_some() {
            eprintln!("pylon-server: CacheInvalidationWorker disabled (--disable-cache-worker)");
        }
    } else if let Some(cache) = cache {
        let dsn = dsn.to_string();
        eprintln!(
            "pylon-server: CacheInvalidationWorker started (shared cache)  channel={}",
            pylon_workers::CACHE_NOTIFY_CHANNEL
        );
        handles.push(tokio::spawn(async move {
            match CacheInvalidationWorker::connect_with_cache(&dsn, cache).await {
                Ok(worker) => worker.run().await,
                Err(e) => eprintln!("pylon-server: failed to start CacheInvalidationWorker: {e}"),
            }
        }));
    } else if toggles.cache == Some(true) {
        eprintln!(
            "pylon-server: --cache-worker given but [cache].enabled = false in pylon.toml; \
             no cache handle to attach to, so there's nothing to invalidate"
        );
    }

    handles
}

fn spawn_index_worker<P>(dsn: String, worker: P, label: &'static str) -> tokio::task::JoinHandle<()>
where
    P: pylon_workers::index_worker::BatchProcessor + 'static,
{
    tokio::spawn(async move {
        if let Err(e) = pylon_workers::index_worker::run(&dsn, BATCH_SIZE, POLL_INTERVAL, worker).await {
            eprintln!("pylon-server: {label} exited: {e}");
        }
    })
}

/// Mirrors `worker.py::_build_providers` — resolves each schema
/// `VectorIndex`'s `[models.*]` entry (falling back to `[models.default]`,
/// like the Python original), logging and skipping any index with no
/// matching config rather than failing the whole worker set.
fn build_providers(schema: &SchemaDescriptor, config: &Config) -> HashMap<(String, Option<String>), ProviderConfig> {
    let mut providers = HashMap::new();
    for t in &schema.types {
        for vi in &t.vector_indexes {
            let type_name = format!("{}::{}", t.module, t.name);
            let Some(model_cfg) = config.models.get(&vi.model).or_else(|| config.models.get("default")) else {
                eprintln!(
                    "pylon-server: no [models.{}] entry in pylon.toml for {type_name} (index={}); skipping",
                    vi.model,
                    vi.index_name.as_deref().unwrap_or("<default>")
                );
                continue;
            };
            providers.insert(
                (type_name, vi.index_name.clone()),
                ProviderConfig {
                    api_style: match model_cfg.api_style {
                        ApiStyle::OpenAi => "openai".to_string(),
                        ApiStyle::Anthropic => "anthropic".to_string(),
                    },
                    api_url: model_cfg.api_url.clone(),
                    model: model_cfg.model.clone(),
                    api_key: model_cfg.secret.clone(),
                },
            );
        }
    }
    providers
}

File: pylon/workers.py
#
# This source file is part of the Pylon open source project.
#
# Copyright (c) 2026 Jaldis B.V.
#
# Licensed under the MIT OR Apache-2.0 license (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://opensource.org/licenses/MIT
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""The background workers that have to run in a Python process.

Two do: the signal dispatcher, because a `@pylon.signal` handler is a live
Python callable that only exists here, and cache invalidation, because it
evicts from an LMDB environment on local disk and so has to reach the cache
some nearby process actually reads. Everything else — vector and search
indexing — claims outbox rows the database arbitrates, can therefore run
anywhere, and runs in `pylon-server`.

`pylon worker start` runs both of these in a process of its own. This module
is the same pair as an importable API, for an application that already has
an event loop and would rather run them beside it than deploy a second
process. Which entry point you want depends on how your framework exposes
the process lifetime: `run_workers` for a single bracketing hook, as an
ASGI lifespan gives you, and `BackgroundWorkers` for a startup callback and
a shutdown callback that don't share a scope.

    @asynccontextmanager
    async def lifespan(app):
        await client.ensure_connected()
        async with run_workers():
            yield

Placement is the whole point for the cache worker. `[cache]` is a file with
no expiry, so an invalidator only keeps a cache correct if it is attached to
*that* cache — by sharing either the process (`shared_cache` below) or the
filesystem path (a separate process on the same machine or volume, which
LMDB supports). A worker with neither is not a partial safeguard: it evicts
from a cache nobody reads, while every reader of the real one keeps serving
the write it never saw, for as long as the process lives.
"""

from __future__ import annotations

import asyncio
import logging
from contextlib import asynccontextmanager
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import AsyncIterator, Collection, Coroutine

    from pylon._core import SchemaDescriptor
    from pylon.config import Config

__all__ = ['WORKER_KINDS', 'BackgroundWorkers', 'build_worker_tasks', 'run_workers']

#: Every worker this module can start, as accepted by the `disabled`
#: argument of `build_worker_tasks`/`run_workers` (and mirrored by
#: `pylon worker start`'s `--disable-*` flags).
WORKER_KINDS = ('cache', 'signals')


def _check_disabled(disabled: Collection[str]) -> frozenset[str]:
    """Normalize and validate a `disabled` argument.

    Rejects an unknown kind rather than ignoring it: a silently-misspelled
    `'caches'` would leave the worker it was meant to suppress running, and
    the whole reason to disable one is that something else is meant to be
    covering it.
    """
    from pylon.exceptions import InterfaceError

    kinds = frozenset(disabled)
    unknown = sorted(kinds - frozenset(WORKER_KINDS))
    if unknown:
        raise InterfaceError(
            f'unknown worker kind(s) {", ".join(repr(k) for k in unknown)}; '
            f'expected any of {", ".join(repr(k) for k in WORKER_KINDS)}'
        )
    return kinds


def _resolve(schema: SchemaDescriptor | None, config: Config | None) -> tuple[SchemaDescriptor, Config]:
    """Fill in whichever of schema/config the caller left out.

    Note the ordering this implies for an application that also holds a
    `Client`: connecting replaces the process-level schema singleton with
    the one the database was last migrated to (see
    `client._install_migrated_schema`). Resolving the schema here *after*
    `ensure_connected()` therefore hands the workers the same migrated
    schema the application's own queries compile against, rather than
    whatever the local `.py` files currently declare.
    """
    if config is None:
        from pylon.config import load_config

        config = load_config()
    if schema is None:
        import pylon
        from pylon.query import _get_schema

        pylon.finalize()
        schema = _get_schema()
    return schema, config


def _warn_unclaimed_indexes(schema: SchemaDescriptor, log: logging.Logger) -> None:
    """Note any index whose outbox rows nothing in this process will drain.

    An unindexed row is invisible: the write succeeds, the outbox row lands,
    and the only symptom is a search result that never appears. Since the
    worker that would drain it lives in a different binary entirely, say so
    once at startup rather than leaving it to be discovered from a query
    returning less than it should.
    """
    kinds = set()
    for td in schema.types:
        if td.vector_indexes:
            kinds.add('VectorIndex')
        if td.search_indexes:
            kinds.add('SearchIndex')
    if kinds:
        log.info(
            '%s declared in the schema — those outboxes are drained by pylon-server, not here',
            '/'.join(sorted(kinds)),
        )


def build_worker_tasks(
    schema: SchemaDescriptor | None = None,
    config: Config | None = None,
    *,
    batch_size: int = 50,
    poll_interval: float = 30.0,
    log: logging.Logger | None = None,
    shared_cache: bool = False,
    disabled: Collection[str] = (),
) -> list[Coroutine[Any, Any, None]]:
    """Build the worker coroutines `schema`/`config` imply, without running them.

    Prefer `run_workers` unless you need to own the tasks yourself — these
    coroutines must be scheduled by the caller, and a worker that stops
    being polled is indistinguishable from one that was never started.

    Args:
        schema: Defaults to the process schema singleton (see `_resolve`).
        config: Defaults to the `pylon.toml` found from the working tree.
        batch_size: Signal-outbox rows claimed per polling cycle.
        poll_interval: Seconds between polls when the outbox is empty.
        log: Where the per-worker startup lines go.
        shared_cache: True when this process has already opened the LMDB
            cache (any process holding a connected `Client` has). The
            cache-invalidation worker then attaches to that handle instead
            of opening a second one, which LMDB refuses within a single
            process. Leave False in a process with no cache of its own —
            the worker opens `[cache].path` itself, and its evictions are
            visible to every other process mapping the same file.
        disabled: Worker kinds to skip, from `WORKER_KINDS` — for a
            deployment where something else already runs them.
    """
    log = log or logging.getLogger(__name__)
    skip = _check_disabled(disabled)
    schema, config = _resolve(schema, config)

    want_cache = config.cache.enabled and 'cache' not in skip
    if want_cache and shared_cache:
        # Checked up front, before a single coroutine exists: raising from
        # the middle of the build would leave any worker already
        # constructed unawaited, which surfaces as a "coroutine was never
        # awaited" warning pointing at this module rather than at the
        # calling-order mistake that caused it.
        from pylon.cache import is_initialized

        if not is_initialized():
            from pylon.exceptions import InterfaceError

            raise InterfaceError(
                'shared_cache=True requires this process to have opened the cache already, '
                'which happens when a Client connects — await Client.ensure_connected() '
                'before starting the workers, or pass shared_cache=False to have the worker '
                'open [cache].path itself.'
            )

    if 'signals' not in skip:
        from pylon.schema._registry import signals_snapshot

        want_signals = bool(signals_snapshot())
    else:
        want_signals = False

    db = config.database
    dsn = db.dsn or f'postgresql://{db.user}:{db.password}@{db.host}:{db.port}/{db.name}'

    _warn_unclaimed_indexes(schema, log)

    tasks: list = []

    if want_cache and shared_cache:
        from pylon._core import run_cache_invalidation_worker_shared
        from pylon.cache import NOTIFY_CHANNEL

        # This process already opened the LMDB handle for its own
        # read-through cache — attach to that same handle rather than
        # opening a second one (LMDB refuses a second Env::open on the same
        # path within one process).
        log.info('CacheInvalidationWorker started (shared cache)  channel=%s', NOTIFY_CHANNEL)
        tasks.append(run_cache_invalidation_worker_shared(dsn))
    elif want_cache:
        from pylon._core import run_cache_invalidation_worker
        from pylon.cache import NOTIFY_CHANNEL

        # Opens its own LMDB handle onto the file-backed cache at
        # config.cache.path. LMDB supports safe concurrent multi-process
        # access to one file, so this process evicting entries is
        # immediately visible to every process mapping the same path.
        log.info('CacheInvalidationWorker started  channel=%s', NOTIFY_CHANNEL)
        tasks.append(run_cache_invalidation_worker(dsn, str(config.cache.path), config.cache.max_size_mb))

    if want_signals:
        from pylon.signals import run_signal_dispatcher

        # Python, unlike every other worker, because it has to hold a live
        # reference to each registered `@pylon.signal` handler — which only
        # exists in this process.
        log.info(
            'Signal dispatcher started  batch_size=%d  poll_interval=%.0fs',
            batch_size,
            poll_interval,
        )
        tasks.append(run_signal_dispatcher(dsn, batch_size=batch_size, poll_interval=poll_interval))

    return tasks


class BackgroundWorkers:
    """A started/stopped handle on the workers, for a framework whose startup
    and shutdown are two separate callbacks.

    Plenty of frameworks expose the process lifetime as a pair of hooks
    rather than as one bracket around it, and a pair has nowhere for an
    `async with` to suspend. Hold one of these instead:

        workers = BackgroundWorkers()

        @app.on_startup
        async def _start():
            await client.ensure_connected()
            await workers.start()

        @app.on_shutdown
        async def _stop():
            await workers.stop()

    Arguments are `build_worker_tasks`'s, except that `shared_cache`
    defaults to True — a process running its workers inline is one that
    holds a `Client`, and therefore one that already has the cache open.

    Where the framework does give you a single bracketing hook, prefer
    `run_workers`, which is this class with the pairing already done.
    """

    def __init__(
        self,
        schema: SchemaDescriptor | None = None,
        config: Config | None = None,
        *,
        batch_size: int = 50,
        poll_interval: float = 30.0,
        log: logging.Logger | None = None,
        shared_cache: bool = True,
        disabled: Collection[str] = (),
    ) -> None:
        self._kwargs = {
            'schema': schema,
            'config': config,
            'batch_size': batch_size,
            'poll_interval': poll_interval,
            'log': log,
            'shared_cache': shared_cache,
            'disabled': disabled,
        }
        self._tasks: list[asyncio.Task] = []

    @property
    def tasks(self) -> list[asyncio.Task]:
        """The running tasks — empty before `start`, and again after `stop`."""
        return list(self._tasks)

    async def start(self) -> None:
        """Build the workers and schedule them on the running loop.

        Async, and deliberately so: the workers must be bound to the loop
        that will actually serve, and requiring a running one here is what
        makes binding them to some other loop impossible rather than merely
        unlikely.

        Nothing is scheduled if the configuration implies no workers, which
        is not an error — an application with no cache and no signal
        handlers has nothing for this to run, and should not have to know
        that to call it.
        """
        if self._tasks:
            from pylon.exceptions import InterfaceError

            raise InterfaceError('workers are already started; call stop() before starting them again')
        # Deliberately not resolved in __init__: schema resolution has to
        # happen after the client connects (see `_resolve`), and an instance
        # built at import time would have captured the pre-migration schema.
        coros = build_worker_tasks(**self._kwargs)
        self._tasks = [asyncio.ensure_future(c) for c in coros]

    async def stop(self) -> None:
        """Cancel the workers and wait for them to finish.

        A no-op if they were never started, so a shutdown hook doesn't have
        to guard against a startup hook that failed before reaching
        `start`. A worker that ended in an exception rather than in
        cancellation re-raises here — one that died silently at startup is
        a cache that quietly stopped being invalidated, and the shutdown
        path is the last chance anyone has to hear about it.
        """
        tasks, self._tasks = self._tasks, []
        if not tasks:
            return
        for task in tasks:
            task.cancel()
        results = await asyncio.gather(*tasks, return_exceptions=True)
        failures = [r for r in results if isinstance(r, BaseException) and not isinstance(r, asyncio.CancelledError)]
        if failures:
            raise failures[0]


@asynccontextmanager
async def run_workers(
    schema: SchemaDescriptor | None = None,
    config: Config | None = None,
    *,
    batch_size: int = 50,
    poll_interval: float = 30.0,
    log: logging.Logger | None = None,
    shared_cache: bool = True,
    disabled: Collection[str] = (),
) -> AsyncIterator[list[asyncio.Task]]:
    """Run the background workers for the duration of the `async with` block.

    Shaped for an ASGI lifespan, which brackets the process lifetime in a
    single hook:

        @asynccontextmanager
        async def lifespan(app):
            await client.ensure_connected()
            async with run_workers():
                yield

        app = FastAPI(lifespan=lifespan)

    Arguments are `BackgroundWorkers`'s, which is what this wraps. Where a
    framework splits startup and shutdown into two separate callbacks
    instead, there is nothing here for the `yield` to suspend inside — hold
    a `BackgroundWorkers` across the pair.

    Yields the started tasks, for a caller that wants to inspect them.
    """
    workers = BackgroundWorkers(
        schema,
        config,
        batch_size=batch_size,
        poll_interval=poll_interval,
        log=log,
        shared_cache=shared_cache,
        disabled=disabled,
    )
    await workers.start()
    try:
        yield workers.tasks
    finally:
        await workers.stop()
